File: problem_021.c
// -*- compile-command: "gcc std=c99 -o problem_021 problem_021.c ceuler.c -Wall -lmxs" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include "ceuler.h"
#include <assert.h>

int main()
{
    printf("Project Euler - Problem 21:\n"
           "Evaluate the sum of all the amicable numbers under 10000.\n\n");

    // Begin time tracking
    struct timeval start;
    gettimeofday(&start, NULL);

    int divisorSum = 0;
    int sum        = 0;
    short history[10001] = {0};
    bool debug = false;

    for (int number=1 ; number<10001 ; number++)
    {
        divisorSum=0;

        // Find sum of divisors
        for (int divisor=1 ; divisor<number/2+1 ; divisor++)
            if (number % divisor == 0)
                divisorSum += divisor;

        // Ignore sums larger than 10000
        if ( divisorSum > 10000 )
            continue;

        // Check if amicable
        if (history[divisorSum] == number)
        {
            if (debug == true) printf("Amicable pair %d %d\n", number, divisorSum); // DEBUG
            sum += number + divisorSum;
        }

        // Record divisor sum
        history[number] = divisorSum;
    }

    printf("Sum: %d\n", sum);
    printElapsedTime(start);

    return 0;
}

File: problem_021.rs
// -*- compile-command: "rustc -o problem_021_rs problem_021.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn main() {
    println!("Project Euler - Problem 21:");
    println!("Evaluate the sum of all the amicable numbers under 10000.\n");

    let mut divisor_sum_table: [usize; 10000] = [0; 10000];

    for i in 1..divisor_sum_table.len() {
        let mut t = 2;
        while t * i < divisor_sum_table.len() {
            divisor_sum_table[t*i] += i as usize;
            t += 1;
        }
    }

    let mut amicable_sum = 0;

    for i in 1..divisor_sum_table.len() {
        let t = divisor_sum_table[i];
        if t < divisor_sum_table.len() && i != t{
            if i == divisor_sum_table[t] {
                amicable_sum += i;
            }
        }
    }

    println!("{}", amicable_sum);
}
